swarm/kubernetes.go

```diff
@@ -0,0 +1,262 @@
+package swarm
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// DefaultNamespace is the default namespace where swarm searches for peer information
+	DefaultNamespace = "kube-system"
+	// DefaultLabelSelectorKey is the default label key to select Pods for peer information
+	DefaultLabelSelectorKey = "application"
+	// DefaultLabelSelectorValue is the default label value to select Pods for peer information
+	DefaultLabelSelectorValue = "skipper-ingress"
+
+	defaultKubernetesURL    = "http://localhost:8001"
+	endpointURIFmt          = "/api/v1/namespaces/%s/endpoints/%s"
+	serviceAccountDir       = "/var/run/secrets/kubernetes.io/serviceaccount/"
+	serviceAccountTokenKey  = "token"
+	serviceAccountRootCAKey = "ca.crt"
+	serviceHostEnvVar       = "KUBERNETES_SERVICE_HOST"
+	servicePortEnvVar       = "KUBERNETES_SERVICE_PORT"
+)
+
+var (
+	errAPIServerURLNotFound = errors.New("kubernetes API server URL could not be constructed from env vars")
+	errInvalidCertificate   = errors.New("invalid CA")
+	errEndpointNotFound     = errors.New("endpoint not found")
+)
+
+// KubernetesOptions are Kubernetes specific swarm options, that are
+// needed to find peers.
+type KubernetesOptions struct {
+	KubernetesInCluster  bool
+	KubernetesAPIBaseURL string
+	Namespace            string
+	LabelSelectorKey     string
+	LabelSelectorValue   string
+}
+
+// ClientKubernetes is the client to access kubernetes resources to find the
+// peers to join a swarm.
+type ClientKubernetes struct {
+	httpClient *http.Client
+	apiURL     string
+	token      string
+}
+
+// Get does the http GET call to kubernetes API to find the initial
+// peers of a swarm.
+func (c *ClientKubernetes) Get(s string) (*http.Response, error) {
+	req, err := c.createRequest("GET", s, nil)
```

Comment:
also could be just `c.httpClient.Get(s)`

Reply:
No because of the potentially used token. We read the token from the secret of the serviceAccount we run.
Maybe I missunderstand your point?

Reply:
true, my bad, missed the token part.